File: appium-tests/reporter.py
import os
import datetime
from openpyxl import Workbook
from openpyxl.styles import Font, PatternFill, Alignment, Border, Side
from openpyxl.utils import get_column_letter


class AppiumExcelReporter:
    def __init__(self, output_path="appium_test_report.xlsx"):
        self.output_path = output_path
        self.steps = []
        self.start_time = datetime.datetime.now()

    def add_step(self, test_case_id, module, description, status,
                 duration_ms, details="", screenshot_path=None):
        """
        Record a test step result.
        :param test_case_id:   e.g. 'TC_001'
        :param module:         e.g. 'Authentication'
        :param description:    Human-readable test description
        :param status:         'PASS' or 'FAIL'
        :param duration_ms:    Execution duration in milliseconds
        :param details:        Extra log context or error message
        :param screenshot_path: Relative path to screenshot file
        """
        self.steps.append({
            "timestamp":      datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
            "test_case_id":   test_case_id,
            "module":         module,
            "description":    description,
            "status":         status.upper(),
            "duration_ms":    duration_ms,
            "details":        details,
            "screenshot":     screenshot_path or "",
        })

    def generate_report(self):
        """Compile and write results into a professionally styled Excel spreadsheet."""
        wb = Workbook()

        # ── SHEETS ──────────────────────────────────────────────────────────────
        ws_summary = wb.active
        ws_summary.title = "Summary Dashboard"
        ws_details = wb.create_sheet(title="Test Details")

        # ── PALETTE & STYLES ────────────────────────────────────────────────────
        FF = "Segoe UI"

        font_title   = Font(name=FF, size=16, bold=True,  color="1F4E79")
        font_section = Font(name=FF, size=12, bold=True,  color="2C3E50")
        font_header  = Font(name=FF, size=11, bold=True,  color="FFFFFF")
        font_bold    = Font(name=FF, size=10, bold=True)
        font_reg     = Font(name=FF, size=10)
        font_detail  = Font(name=FF, size=9,  italic=True, color="555555")

        fill_header  = PatternFill("solid", fgColor="1F4E79")
        fill_zebra   = PatternFill("solid", fgColor="F2F7FA")
        fill_pass    = PatternFill("solid", fgColor="D4EDDA")
        fill_fail    = PatternFill("solid", fgColor="F8D7DA")
        fill_accent  = PatternFill("solid", fgColor="E9ECEF")

        thin  = Side(border_style="thin",   color="D3D3D3")
        thick = Side(border_style="medium", color="1F4E79")
        b_all = Border(left=thin, right=thin, top=thin, bottom=thin)
        b_hdr = Border(left=thin, right=thin, top=thin, bottom=thick)

        ctr  = Alignment(horizontal="center", vertical="center")
        lft  = Alignment(horizontal="left",   vertical="center")
        rgt  = Alignment(horizontal="right",  vertical="center")

        # ── SUMMARY DASHBOARD ───────────────────────────────────────────────────
        end_time    = datetime.datetime.now()
        total_dur   = end_time - self.start_time
        total       = len(self.steps)
        passed      = sum(1 for s in self.steps if s["status"] == "PASS")
        failed      = total - passed
        pass_rate   = (passed / total * 100) if total > 0 else 0.0

        ws_summary.merge_cells("A1:E1")
        ws_summary["A1"] = "SkillSphereAI — Appium Mobile E2E Test Report"
        ws_summary["A1"].font = font_title
        ws_summary["A1"].alignment = lft
        ws_summary.row_dimensions[1].height = 42

        ws_summary["A3"] = "Execution Summary & KPIs"
        ws_summary["A3"].font = font_section
        ws_summary.row_dimensions[3].height = 26

        metrics = [
            ("Execution Start",   self.start_time.strftime("%Y-%m-%d %H:%M:%S")),
            ("Execution End",     end_time.strftime("%Y-%m-%d %H:%M:%S")),
            ("Total Duration",    f"{total_dur.total_seconds():.2f} seconds"),
            ("Total Test Cases",  total),
            ("Cases Passed",      passed),
            ("Cases Failed",      failed),
            ("Pass Rate",         f"{pass_rate:.1f}%"),
        ]

        row = 4
        for label, value in metrics:
            lc = ws_summary.cell(row=row, column=1, value=label)
            lc.font = font_bold; lc.alignment = lft
            lc.fill = fill_accent; lc.border = b_all

            vc = ws_summary.cell(row=row, column=2, value=value)
            vc.font = font_reg; vc.alignment = lft; vc.border = b_all

            if label == "Pass Rate":
                vc.font = Font(name=FF, size=10, bold=True,
                               color="2E7D32" if pass_rate >= 80 else "C62828")
                vc.fill = fill_pass if pass_rate >= 80 else fill_fail
            elif label == "Cases Failed" and failed > 0:
                vc.font = Font(name=FF, size=10, bold=True, color="C62828")
                vc.fill = fill_fail
            elif label == "Cases Passed" and passed > 0:
                vc.font = Font(name=FF, size=10, bold=True, color="2E7D32")

            ws_summary.row_dimensions[row].height = 22
            row += 1

        ws_summary.column_dimensions["A"].width = 22
        ws_summary.column_dimensions["B"].width = 30

        # ── TEST DETAILS SHEET ──────────────────────────────────────────────────
        headers = ["No.", "TC ID", "Module", "Test Description",
                   "Status", "Duration (ms)", "Execution Log", "Screenshot"]

        for ci, h in enumerate(headers, 1):
            c = ws_details.cell(row=1, column=ci, value=h)
            c.font = font_header; c.fill = fill_header
            c.alignment = ctr; c.border = b_hdr
        ws_details.row_dimensions[1].height = 28

        for ri, step in enumerate(self.steps, 2):
            is_even = ri % 2 == 0
            row_bg  = fill_zebra if is_even else PatternFill()

            def cell(col, val, align=lft):
                c = ws_details.cell(row=ri, column=col, value=val)
                c.alignment = align; c.border = b_all
                if step["status"] not in ("PASS", "FAIL"):
                    c.fill = row_bg
                return c

            cell(1, ri - 1, ctr)
            cell(2, step["test_case_id"], ctr)
            cell(3, step["module"])
            cell(4, step["description"])

            sc = ws_details.cell(row=ri, column=5, value=step["status"])
            sc.alignment = ctr; sc.border = b_all
            if step["status"] == "PASS":
                sc.fill = fill_pass
                sc.font = Font(name=FF, size=10, bold=True, color="155724")
            else:
                sc.fill = fill_fail
                sc.font = Font(name=FF, size=10, bold=True, color="721C24")

            cell(6, step["duration_ms"], rgt)

            dc = ws_details.cell(row=ri, column=7, value=step["details"])
            dc.alignment = lft; dc.border = b_all
            dc.font = (Font(name=FF, size=9, bold=True, color="721C24")
                       if step["status"] == "FAIL" else font_detail)

            sc2 = ws_details.cell(row=ri, column=8, value=step["screenshot"])
            sc2.alignment = lft; sc2.border = b_all; sc2.font = font_detail

            ws_details.row_dimensions[ri].height = 20

        # Auto-fit columns
        col_widths = [6, 10, 24, 52, 10, 14, 55, 35]
        for i, w in enumerate(col_widths, 1):
            ws_details.column_dimensions[get_column_letter(i)].width = w

        wb.save(self.output_path)
        print(f"\n[+] Excel report saved -> {os.path.abspath(self.output_path)}")


if __name__ == "__main__":
    r = AppiumExcelReporter("self_test_report.xlsx")
    r.add_step("TC_001", "Authentication",  "App launches and auth panel renders",       "PASS", 850,  "Auth screen detected.")
    r.add_step("TC_002", "Authentication",  "Email field is visible",                    "PASS", 320,  "EditText[1] found.")
    r.add_step("TC_006", "Authentication",  "Valid email entered",                       "PASS", 480,  "tony6250584@gmail.com typed.")
    r.add_step("TC_013", "Onboarding",      "Demographics screen visible",               "PASS", 600,  "Onboarding container rendered.")
    r.add_step("TC_066", "Scorecard",       "Scorecard tab navigated",                   "FAIL", 1200, "Tab not found within timeout.", "screenshots/tc066_fail.png")
    r.generate_report()

File: selenium-tests/reporter.js
import ExcelJS from "exceljs";
import path from "path";
import fs from "fs";

export class SeleniumExcelReporter {
  constructor(outputPath = "selenium_test_report.xlsx") {
    this.outputPath = outputPath;
    this.steps = [];
    this.startTime = new Date();
  }

  /**
   * Log a test step.
   * @param {string} stepName Name of the action/step
   * @param {string} status 'PASS' or 'FAIL'
   * @param {number} durationMs Execution duration in milliseconds
   * @param {string} details Optional execution details or error logs
   * @param {string} screenshotPath Path to visual screenshot
   */
  addStep(stepName, status, durationMs, details = "", screenshotPath = "") {
    const timestampStr = new Date().toISOString().replace("T", " ").substring(0, 19);
    this.steps.push({
      timestamp: timestampStr,
      stepName,
      status: status.toUpperCase(),
      durationMs,
      details,
      screenshot: screenshotPath
    });
  }

  /**
   * Compiles the recorded steps into a beautifully styled Excel workbook using exceljs.
   */
  async generateReport() {
    const workbook = new ExcelJS.Workbook();
    
    // Create worksheets
    const wsSummary = workbook.addWorksheet("Summary Dashboard", {
      views: [{ showGridLines: true }]
    });
    const wsDetails = workbook.addWorksheet("Test Details", {
      views: [{ showGridLines: true }]
    });

    // Formatting Tokens
    const fontFamily = "Segoe UI";
    
    const styleTitle = { name: fontFamily, size: 16, bold: true, color: { argb: "FF1F4E79" } };
    const styleSection = { name: fontFamily, size: 12, bold: true, color: { argb: "FF2C3E50" } };
    const styleHeader = { name: fontFamily, size: 11, bold: true, color: { argb: "FFFFFFFF" } };
    const styleBold = { name: fontFamily, size: 10, bold: true };
    const styleRegular = { name: fontFamily, size: 10 };
    const styleDetails = { name: fontFamily, size: 9, italic: true, color: { argb: "FF555555" } };
    
    const fillHeader = { type: "pattern", pattern: "solid", fgColor: { argb: "FF1F4E79" } }; // Deep Teal
    const fillZebra = { type: "pattern", pattern: "solid", fgColor: { argb: "FFF2F7FA" } };  // Light Blue-Gray
    const fillPass = { type: "pattern", pattern: "solid", fgColor: { argb: "FFD4EDDA" } };   // Soft Green
    const fillFail = { type: "pattern", pattern: "solid", fgColor: { argb: "FFF8D7DA" } };   // Soft Red
    const fillAccent = { type: "pattern", pattern: "solid", fgColor: { argb: "FFE9ECEF" } }; // Neutral Gray Accent
    
    const borderThinSide = { style: "thin", color: { argb: "FFD3D3D3" } };
    const borderAll = {
      top: borderThinSide,
      left: borderThinSide,
      bottom: borderThinSide,
      right: borderThinSide
    };
    
    const borderHeader = {
      top: borderThinSide,
      left: borderThinSide,
      bottom: { style: "medium", color: { argb: "FF1F4E79" } },
      right: borderThinSide
    };

    const alignCenter = { horizontal: "center", vertical: "middle" };
    const alignLeft = { horizontal: "left", vertical: "middle" };
    const alignRight = { horizontal: "right", vertical: "middle" };

    // --- 1. BUILD SUMMARY DASHBOARD ---
    const endTime = new Date();
    const totalTimeSec = ((endTime.getTime() - this.startTime.getTime()) / 1000).toFixed(2);
    const totalSteps = this.steps.length;
    const passedSteps = this.steps.filter(s => s.status === "PASS").length;
    const failedSteps = this.steps.filter(s => s.status === "FAIL").length;
    const passRate = totalSteps > 0 ? ((passedSteps / totalSteps) * 100).toFixed(1) : "0.0";

    // Title Row
    wsSummary.mergeCells("A1:D1");
    const titleCell = wsSummary.getCell("A1");
    titleCell.value = "SkillSphereAI Web E2E Test Report";
    titleCell.font = styleTitle;
    titleCell.alignment = alignLeft;
    wsSummary.getRow(1).height = 40;

    wsSummary.getRow(2).height = 10; // spacing row

    // Section header
    const sectionCell = wsSummary.getCell("A3");
    sectionCell.value = "Test Summary & Performance Indicators";
    sectionCell.font = styleSection;
    wsSummary.getRow(3).height = 25;

    // Metrics Data
    const metrics = [
      ["Execution Start", this.startTime.toISOString().replace("T", " ").substring(0, 19)],
      ["Execution End", endTime.toISOString().replace("T", " ").substring(0, 19)],
      ["Total Duration", `${totalTimeSec} seconds`],
      ["Total Test Steps", totalSteps],
      ["Steps Passed", passedSteps],
      ["Steps Failed", failedSteps],
      ["Pass Rate", `${passRate}%`]
    ];

    let summaryRow = 4;
    for (const [metricName, val] of metrics) {
      const nameCell = wsSummary.getCell(`A${summaryRow}`);
      nameCell.value = metricName;
      nameCell.font = styleBold;
      nameCell.alignment = alignLeft;
      nameCell.fill = fillAccent;
      nameCell.border = borderAll;

      const valCell = wsSummary.getCell(`B${summaryRow}`);
      valCell.value = val;
      valCell.font = styleRegular;
      valCell.alignment = alignLeft;
      valCell.border = borderAll;

      // Conditional formatting for summary indicators
      if (metricName === "Pass Rate") {
        valCell.font = { name: fontFamily, size: 10, bold: true, color: { argb: parseFloat(passRate) >= 80 ? "FF2E7D32" : "FFC62828" } };
        valCell.fill = parseFloat(passRate) >= 80 ? fillPass : fillFail;
      } else if (metricName === "Steps Failed" && failedSteps > 0) {
        valCell.font = { name: fontFamily, size: 10, bold: true, color: { argb: "FFC62828" } };
        valCell.fill = fillFail;
      } else if (metricName === "Steps Passed" && passedSteps > 0) {
        valCell.font = { name: fontFamily, size: 10, bold: true, color: { argb: "FF2E7D32" } };
      }

      wsSummary.getRow(summaryRow).height = 20;
      summaryRow++;
    }

    // --- 2. BUILD DETAILS SHEET ---
    const headers = ["No.", "Timestamp", "Step Name", "Status", "Duration (ms)", "Execution Details", "Screenshot Path"];
    const headerRow = wsDetails.getRow(1);
    headerRow.height = 26;

    for (let i = 0; i < headers.length; i++) {
      const cell = headerRow.getCell(i + 1);
      cell.value = headers[i];
      cell.font = styleHeader;
      cell.fill = fillHeader;
      cell.alignment = alignCenter;
      cell.border = borderHeader;
    }

    let detailRowIdx = 2;
    for (const step of this.steps) {
      const r = wsDetails.getRow(detailRowIdx);
      r.height = 20;

      r.getCell(1).value = detailRowIdx - 1;
      r.getCell(1).alignment = alignCenter;

      r.getCell(2).value = step.timestamp;
      r.getCell(2).alignment = alignCenter;

      r.getCell(3).value = step.stepName;
      r.getCell(3).alignment = alignLeft;

      const statusCell = r.getCell(4);
      statusCell.value = step.status;
      statusCell.alignment = alignCenter;
      if (step.status === "PASS") {
        statusCell.fill = fillPass;
        statusCell.font = { name: fontFamily, size: 10, bold: true, color: { argb: "FF155724" } };
      } else {
        statusCell.fill = fillFail;
        statusCell.font = { name: fontFamily, size: 10, bold: true, color: { argb: "FF721C24" } };
      }

      r.getCell(5).value = step.durationMs;
      r.getCell(5).alignment = alignRight;

      const detailsCell = r.getCell(6);
      detailsCell.value = step.details;
      detailsCell.alignment = alignLeft;
      if (step.status === "FAIL") {
        detailsCell.font = { name: fontFamily, size: 9, bold: true, color: { argb: "FF721C24" } };
      } else {
        detailsCell.font = styleDetails;
      }

      r.getCell(7).value = step.screenshot;
      r.getCell(7).alignment = alignLeft;
      r.getCell(7).font = styleDetails;

      // Apply borders & zebra striping
      for (let c = 1; c <= 7; c++) {
        const cell = r.getCell(c);
        cell.border = borderAll;
        if (detailRowIdx % 2 === 1 && step.status !== "PASS" && step.status !== "FAIL") {
          cell.fill = fillZebra;
        }
      }

      detailRowIdx++;
    }

    // --- AUTO-FIT COLUMNS ---
    wsSummary.columns.forEach(col => {
      let maxLen = 0;
      col.eachCell({ includeEmpty: false }, cell => {
        const valStr = cell.value ? String(cell.value) : "";
        if (valStr.length > maxLen) {
          maxLen = valStr.length;
        }
      });
      col.width = Math.max(maxLen + 3, 12);
    });

    wsDetails.columns.forEach(col => {
      let maxLen = 0;
      col.eachCell({ includeEmpty: false }, cell => {
        const valStr = cell.value ? String(cell.value) : "";
        if (valStr.length > maxLen) {
          maxLen = valStr.length;
        }
      });
      col.width = Math.max(maxLen + 3, 12);
    });

    // Custom overrides for widths
    wsSummary.getColumn(1).width = 25;
    wsSummary.getColumn(2).width = 30;
    wsDetails.getColumn(6).width = 50;
    wsDetails.getColumn(7).width = 40;

    await workbook.xlsx.writeFile(this.outputPath);
    console.log(`Excel test analysis report compiled: ${path.resolve(this.outputPath)}`);
  }
}

// Self-test execution guard
if (process.argv[1] && process.argv[1].endsWith("reporter.js")) {
  (async () => {
    const reporter = new SeleniumExcelReporter("self_test_report.xlsx");
    reporter.addStep("Launch Browser", "PASS", 740, "Chrome browser launched and pointed to homepage.");
    reporter.addStep("Enter Logins", "PASS", 1100, "Successfully authenticated as Candidate User.");
    reporter.addStep("Run Assessment", "FAIL", 1450, "Assertion Error: Scores didn't increment by 10 points on client dashboard.", "screenshots/fail_step_3.png");
    await reporter.generateReport();
    if (fs.existsSync("self_test_report.xlsx")) {
      fs.unlinkSync("self_test_report.xlsx");
      console.log("Self test report verified and cleaned up.");
    }
  })();
}
